src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file provides a utility to generate Additional Data payload and its members
+ *      (e.g. rotating device id)
+ *
+ */
+
+#include "AdditionalDataPayloadGenerator.h"
+#include "AdditionalDataPayload.h"
+
+#include <core/CHIPCore.h>
+#include <core/CHIPEncoding.h>
+#include <core/CHIPSafeCasts.h>
+#include <core/CHIPTLV.h>
+#include <crypto/CHIPCryptoPAL.h>
+#include <stdlib.h>
+#include <support/BufferWriter.h>
+#include <support/CHIPMem.h>
+
+using namespace chip;
+using namespace chip::System;
+using namespace chip::TLV;
+using namespace chip::Crypto;
+using namespace chip::SetupPayloadData;
+using namespace chip::Encoding::LittleEndian;
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(uint16_t lifetimeCounter, const char * serialNumberBuffer,
+                                                                         size_t serialNumberBufferSize,
+                                                                         PacketBufferHandle & bufferHandle,
+                                                                         AdditionalDataFields additionalDataFields)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[RotatingDeviceId::kRotatingDeviceIdHexMaxLength];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize));
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if ((additionalDataFields & AdditionalDataFields::RotatingDeviceId) == AdditionalDataFields::RotatingDeviceId)
```

Comment:
Just `!= 0` would work too, since it's a single bit, right?  Might be clearer.

Reply:
updated